### ISSUE TITLE
db.transact() usage with tokio

### ISSUE THREAD

**Issue by coadler** [Clikengo/foundationdb-rs#158]
I don't seem to be able to use `db.transact()` within `tokio`.

Consider the following program:

```toml
[dependencies]
tokio = { version = "0.2.4", features = ["full"] }
foundationdb = { git = "https://github.com/bluejekyll/foundationdb-rs" }
futures = "0.3.1"
```

```rust
extern crate foundationdb;
extern crate tokio;

use foundationdb::*;
use futures::future::*;
use std::error::Error;
use std::sync::Arc;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    foundationdb::boot().expect("failed to boot network");

    do_transact().await;
    do_trx().await;

    Ok(())
}

async fn do_transact() {
    let db = Arc::new(foundationdb::Database::default().expect("failed to open fdb"));

    let adb = db.clone();
    tokio::spawn(async move {
        async fn txnFn(txn: &Transaction) -> FdbResult<()> {
            Ok(())
        }

        adb.transact(
            (),
            |txn: &Transaction, ()| txnFn(txn).boxed_local(),
            TransactOption::default(),
        )
        .await
        .expect("failed to transact")
    });
}

async fn do_trx() {
    let db = Arc::new(foundationdb::Database::default().expect("failed to open fdb"));

    let adb = db.clone();
    tokio::spawn(async move {
        adb.create_trx()
            .expect("failed to create trx")
            .commit()
            .await
            .expect("failed to commit");
    });
}
```

`tokio::spawn` inside `do_transact` throws an error, while `do_trx` is fine.

```
$ cargo build
   Compiling fdb-test v0.1.0 (/home/colin/rust/src/coadler/fdb-test)
error[E0277]: `dyn core::future::future::Future<Output = std::result::Result<(), foundationdb::FdbError>>` cannot be sent between threads safely
   --> src/main.rs:23:5
    |
23  |     tokio::spawn(async move {
    |     ^^^^^^^^^^^^ `dyn core::future::future::Future<Output = std::result::Result<(), foundationdb::FdbError>>` cannot be sent between threads safely
    |
   ::: /home/colin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.4/src/task/spawn.rs:123:21
    |
123 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
```

I'm not sure if this is expected behavior on master.

cc @Speedy37

**Comment by Speedy37** [comment]
You are right, db.transact should probably be return a Future + Send + Sync. I'll look into it.

**Comment by Speedy37** [comment]
This is caused by F being a Box: 

```rust
Pin<Box<dyn Future<Output = Result<Item, Error>> + Send + Sync + 'a>>
```

versus

```rust
Pin<Box<dyn Future<Output = Result<Item, Error>> + 'a>>
```

I will do two variants then: transact and transact_local (like boxed, boxed_local).

Ideally I hope GAT will let me remove the Pin<Box<dyn Future + 'a>> thing, but for that particular case, GAT would probably not simplify things. I should build a nightly version of transact (behind the gat feature flag to showcase what I hope to achieve once GAT hit stable).

**Comment by Speedy37** [comment]
I did some investigations:
- If instead of Pin<Box<dyn Future>>, the real type is known, then the compiler is able to auto impl the Send trait.
- I was unable to type the transact method in a way that is both easy to use and let the compiler access the concrete type

**Comment by Speedy37** [comment]
Fixed by #162
